File: CountAndSay.java

public class CountAndSay {
    public static String countAndSay(int n) {
        StringBuilder sb = new StringBuilder();
        sb.append("1");
        while (n > 1) {
            String temp = sb.toString();
            sb = new StringBuilder();
            for (int i = 0; i < temp.length(); ) {
                char ch = temp.charAt(i);
                int count = 1;
                while (i + count < temp.length() && ch == temp.charAt(i + count))
                    count++;
                sb.append(String.valueOf(count));
                sb.append(ch);
                i += count;
            }    
            n--;
            
        }
        return sb.toString();
    }
    public static void main(String[] args) {
        System.out.println(countAndSay(1));
        System.out.println(countAndSay(2));
        System.out.println(countAndSay(3));
        System.out.println(countAndSay(4));
        
    }
}

File: py_solution/count_and_say.py
class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        res = '1'
        while n > 1:
            temp = []
            count = 1
            cur = res[0]
            for i in range(1, len(res)):
                if res[i] == cur:
                    count += 1
                else:
                    temp.append(str(count))
                    temp.append(cur)
                    cur = res[i]
                    count = 1
            if count:
                temp.append(str(count))
                temp.append(cur)
            res = ''.join(temp)
            n -= 1
        return res


if __name__ == "__main__":
    solution = Solution()
    # assert solution.countAndSay(1) == '1'
    assert solution.countAndSay(4) == '1211'